shotover-proxy/src/transforms/cassandra/sink_cluster.rs

```diff
@@ -93,60 +123,185 @@ impl CassandraSinkCluster {
+
+                    // Make sure to drop write_lock before the expensive_drop which may have to perform many deallocations.
+                    std::mem::drop(write_lock);
+                    std::mem::drop(expensive_drop);
+                }
+            }
+        }
+
+        Ok(responses)
+    }
+
+    fn get_system_peers_index(&self, messages: &mut [Message]) -> Option<QueryOccured> {
+        let mut system_peers_index: Option<QueryOccured> = None;
+        for (message_index, message) in messages.iter_mut().enumerate() {
+            if let Some(Frame::Cassandra(cassandra)) = message.frame() {
+                // No need to handle Batch as selects can only occur on Query
+                if let CassandraOperation::Query { query, .. } = &cassandra.operation {
+                    if let CassandraStatement::Select(select) = query.as_ref() {
+                        if select.where_clause.is_empty() && select.table_name == self.peer_table {
+                            let mut alias_to_name = HashMap::new();
+                            for select_element in &select.columns {
+                                match select_element {
+                                    SelectElement::Star => {
+                                        // Always overwrite, give priority to * entries
+                                        return Some(QueryOccured {
+                                            message_index,
+                                            alias_to_name,
+                                        });
+                                    }
+                                    SelectElement::Column(ident) => {
+                                        if let Some(alias) = &ident.alias {
+                                            alias_to_name.insert(alias.clone(), ident.name.clone());
+                                        }
+                                    }
+                                    _ => {}
                                 }
                             }
-                            responses.push(message);
+                            // Give priority to existing entry in the hope that its a * entry
+                            if system_peers_index.is_none() {
+                                system_peers_index = Some(QueryOccured {
+                                    message_index,
+                                    alias_to_name,
+                                })
+                            }
                         }
-                        Response {
-                            mut original,
-                            response: Err(err),
-                        } => {
-                            original.set_error(err.to_string());
-                            responses.push(original);
+                    }
+                }
+            }
+        }
+        system_peers_index
+    }
+}
+
+fn get_nodes_from_system_peers(
```

Comment:
this function (or higher up in the call stack) would be a good candidate for an integration test to pick up changes to the system.peers table.

Reply:
The two straightforward ways I can think of doing an integration test here are:
* call `create_topology_task` pass it a hardcoded handshake and then wait until it writes some nodes to the RwLock, and run some assertions on them
   + doesnt test the full stack
   + least invasive approach
   + lets us make assertions directly on the system.peers table that the task returns.
* Add some some metrics counters to CassandraSinkSingle that record each hit to each outgoing node. Then assert that each outgoing node is hit at least once.
    + Could fail if we got really unlucky, it is unlikely though as our integration tests perform a lot of queries.
    + Can only make a very weak assertion against our routing. e.g. impossible to test that we have implemented token aware routing well.

Which of these sounds better?

Something else thats probably too complicated to invest in:
If we wanted a really robust test we could mock out the cassandra nodes and then use data from the mocks to assert that our token aware routing is implemented properly.

Reply:
Ended up just going with the first option.